launcher_go/v2/config/config.go

```diff
@@ -0,0 +1,209 @@
+	Templates       []string          `yaml:templates,omitempty`
+	Expose          []string          `yaml:expose,omitempty`
+	Params          map[string]string `yaml:params,omitempty`
+	Env             map[string]string `yaml:env,omitempty`
+	Labels          map[string]string `yaml:labels,omitempty`
+	Volumes         []struct {
+		Volume struct {
+			Host  string `yaml:host`
+			Guest string `yaml:guest`
+		} `yaml:volume`
+	} `yaml:volumes,omitempty`
+	Links []struct {
+		Link struct {
+			Name  string `yaml:name`
+			Alias string `yaml:alias`
+		} `yaml:link`
+	} `yaml:links,omitempty`
+}
+
+func (config *Config) loadTemplate(templateDir string, template string) error {
+	template_filename := strings.TrimRight(templateDir, "/") + "/" + string(template)
+	content, err := os.ReadFile(template_filename)
+	if err != nil {
+		if os.IsNotExist(err) {
+			fmt.Println("template file does not exist: " + template_filename)
+		}
+		return err
+	}
+	templateConfig := &Config{}
+	if err := yaml.Unmarshal(content, templateConfig); err != nil {
+		return err
+	}
+	if err := mergo.Merge(config, templateConfig, mergo.WithOverride); err != nil {
+		return err
+	}
+	config.rawYaml = append(config.rawYaml, string(content[:]))
+	return nil
+}
+
+func LoadConfig(dir string, configName string, includeTemplates bool, templatesDir string) (*Config, error) {
+	config := &Config{
+		Name:         configName,
+		Boot_Command: DefaultBootCommand,
+		Base_Image:   DefaultBaseImage(),
+	}
+	matched, _ := regexp.MatchString("[[:upper:]/ !@#$%^&*()+~`=]", configName)
+	if matched {
+		msg := "ERROR: Config name '" + configName + "' must not contain upper case characters, spaces or special characters. Correct config name and rerun."
+		fmt.Println(msg)
+		return nil, errors.New(msg)
+	}
+
+	config_filename := string(strings.TrimRight(dir, "/") + "/" + config.Name + ".yml")
+	content, err := os.ReadFile(config_filename)
+	if err != nil {
+		if os.IsNotExist(err) {
+			fmt.Println("config file does not exist: " + config_filename)
+		}
+		return nil, err
+	}
+	baseConfig := &Config{}
+
+	if err := yaml.Unmarshal(content, baseConfig); err != nil {
+		return nil, err
+	}
+
+	if includeTemplates {
+		for _, t := range baseConfig.Templates {
+			if err := config.loadTemplate(templatesDir, t); err != nil {
+				return nil, err
+			}
+		}
+	}
+	if err := mergo.Merge(config, baseConfig, mergo.WithOverride); err != nil {
+		return nil, err
+	}
+	config.rawYaml = append(config.rawYaml, string(content[:]))
+	if err != nil {
+		return nil, err
+	}
+
+	for k, v := range config.Labels {
+		val := strings.ReplaceAll(v, "{{config}}", config.Name)
+		config.Labels[k] = val
+	}
+
+	for k, v := range config.Env {
+		val := strings.ReplaceAll(v, "{{config}}", config.Name)
+		config.Env[k] = val
+	}
+
+	return config, nil
+}
+
+func (config *Config) Yaml() string {
+	return strings.Join(config.rawYaml, "_FILE_SEPERATOR_")
+}
+
+func (config *Config) Dockerfile(pupsArgs string, bakeEnv bool) string {
```

Comment:
What is the purpose of generating a Dockerfile on the fly and using it to build an image? Am wondering why we made a choice here to move away from the `cidfile` approach in building a Docker image with pups.

Reply:
Is this to support a customizable boot command in a sensible way?

Reply:
O ok I get it now, we want to bake things like `EXPOSE` and `ENV` into the Docker image without having to specify those envs when running the image.

Reply:
I've been paying only partial attention to this, so I may be off base. When I've done something like what I think this is doing, I built my images with launcher/pups to (and added env variables to migrate database and such on startup).

> What is the purpose of generating a Dockerfile on the fly and using it to build an image? Am wondering why we made a choice here to move away from the cidfile approach in building a Docker image with pups.

I think the reason to use Dockerfile is because it's the industry standard. I've always thought that if this project had started a year or two later, there would be no pups because there would be no reason for it. When it started, docker-compose either didnt' exist, or was pretty unusable. It was years and years before it got included by default.

Software is more useful when more people are using it? Who's using pups other than us?

Reply:
IMO it's better compatibility to be able to convert our pups config out to a dockerfile - we can rely on `docker build` directly for builds, and having a dockerfile available allows easier exporting of our pups configs out to other tools. It feels much more natural to be able to build images using `docker build` imo.

Future commits add exporting of config out to docker-compose, which also allow a user to build via a dockerfile.